opentelemetry/proto/common/v1/common.proto

```diff
@@ -21,30 +21,37 @@ option java_package = "io.opentelemetry.proto.common.v1";
 option java_outer_classname = "CommonProto";
 option go_package = "github.com/open-telemetry/opentelemetry-proto/gen/go/common/v1";
 
+// AnyValue is used to represent any type of attribute value. AnyValue may contain a
+// primitive value such as a string or integer or it may contain an arbitrary nested
+// object containing arrays, key-value lists and primitives.
+message AnyValue {
+  // The value is one of the listed fields. It is valid for all values to be unspecified
+  // in which case this AnyValue is considered to be "null".
+  oneof value {
+    string string_value = 1;
+    bool bool_value = 2;
+    int64 int_value = 3;
+    double double_value = 4;
+    ArrayValue array_value = 5;
+    AttributeKeyValueList kvlist_value = 6;
+  }
+}
+
+message ArrayValue {
+  // Array of values. The array may be empty (contain 0 elements).
+  repeated AnyValue values = 1;
```

Comment:
This would allow heterogeneous arrays, whereas the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/1729bc4337bdc1df83ff7eab53eb7405d6e75665/specification/trace/api.md#set-attributes) prescribes:
> The array MUST be homogeneous, i.e. it MUST NOT contain values of different types.

Should this be reflected in the protocol as well or are we fine with having this constraint ensured only by the SDK implementations?

Reply:
I'll advocate for changing the specification. I have never understood why it matters that arrays are homogenous when JSON values are what people use in the real world.

Reply:
The protocol can be a bit more generic than the API. With this we can ensure that we have a path forward and we don't have to break the protocol if that changes.

Reply:
@arminru The protocol does not have to be an exact match of the spec. It needs to be able to efficiently and unambiguously represent what the spec defines. This schema can unambiguously represent homogeneous arrays. It happens to also be able to represent heterogeneous arrays, no harm from that. In fact, we gain ability to represent JSON-like data properly, which is a big benefit IMO.

@jmacd some motivation for homogeneous arrays in spec is here: https://github.com/open-telemetry/opentelemetry-specification/issues/341#issuecomment-559708227
I don't have a strong opinion on this.